### ISSUE TITLE
[Bug][Solved]opponent next action inference in centralized Q critic

### ISSUE THREAD

**Issue by XuehaiPan** [Replicable-MARL/MARLlib#41]
Is the `centralized_critic_q` postprocessing only implemented for MADDPG algorithm?

https://github.com/Replicable-MARL/MARLlib/blob/a0fe513c5c31f4d1e5df5f053537896af7abfd4f/marl/algos/utils/postprocessing.py#L347-L370

To my best knowledge for `RLlib`, the implementation above only works with strict conditions. It only works for `DDPG / TD3` policies with both `normalize_actions` and `clip_actions` set to `False`.

1. `policy.model` doesn't have the attribute `policy_model`.
2. In the sample batch, `SampleBatch.ACTIONS` are the real actions taken while interacting with the environment. For example, for discrete action spaces, `SampleBatch.ACTIONS` are integers, not action logits.
3. Missing action logits to action distribution mapping.
4. Missing action unsquashing (unnormalize actions from `[-1, +1]` to `[low, high]`).

**Comment by XuehaiPan** [comment]
Just a question, for an off-policy algorithm, should we always infer the opponent's next actions with the latest opponent policy?

Why not grab the opponent next actions from the batch by:

```python
batch['opponent_next_action'][:-1] = opponent_batch[SampleBatch.ACTIONS][1:]
```

and only inference the last step:

```python
batch['opponent_next_action'][-1] = inference(opponent_batch[SampleBatch.NEXT_OBS][-1], state=...)
```

This implementation will use the old opponent policy to calculate `opponent_next_action` and never update during training.

**Comment by Theohhhu** [comment]
Thank you for your attention.

Our implementation is based on RLlib with necessary modifications.
Here we address your concern one by one.

**Q: <code>policy.model</code> doesn't have the attribute <code>policy_model</code>.**
A: <code>policy.model.policy_model</code> can be found here https://github.com/Replicable-MARL/MARLlib/blob/a0fe513c5c31f4d1e5df5f053537896af7abfd4f/marl/algos/core/IL/ddpg.py#L367

**Q: In the sample batch, <code>SampleBatch.ACTIONS</code> are the real actions taken while interacting with the environment. For example, for discrete action spaces, <code>SampleBatch.ACTIONS</code> are integers, not action logits.**
A:  We build the RNN version of DDPG/MADDPG based on the RLlib one. not the contributed one by OpenAI (I think?). Thus, our version only supports continuous action and the output of the actor is real action (after squashing).

**Q: Missing action logits to action distribution mapping.**
A: Same as the previous question.

**Q: Missing action unsquashing (unnormalize actions from [-1, +1] to [low, high]).**
A: In the learning process, the range of action is [-1, 1]. We only need action unsquashing when interacting with the environment.

**Q:  for an off-policy algorithm, should we always infer the opponent's next actions with the latest opponent policy? This implementation will use the old opponent policy to calculate opponent_next_action and never update during training.**
A: Thanks for pointing this out, we are aware this is a potential issue in DDPG-based algorithms. We fixed this by moving the 
 <code>opponent_next_action</code> and  <code>opponent_next_q</code> from  <code>centralized_critic_q()</code> function to https://github.com/Replicable-MARL/MARLlib/blob/015dd257ccf196ba9cf59385f563e352f1a8ea75/marl/algos/core/CC/maddpg.py#L363
In this way, the <code>opponent_next_action</code> is predicted by the latest <code>policy.target_model</code> right before entering the training iteration.
We also do a similar fix to FACMAC.

Any further concern is welcome.

**Comment by XuehaiPan** [comment]
Refer to my first question at the top:

>> Is the centralized_critic_q postprocessing only implemented for MADDPG algorithm?

> **Q: `policy.model` doesn't have the attribute `policy_model`.**
> A: `policy.model.policy_model` can be found here
> 
> https://github.com/Replicable-MARL/MARLlib/blob/a0fe513c5c31f4d1e5df5f053537896af7abfd4f/marl/algos/core/IL/ddpg.py#L367

According to the answer, only MADDPG and its variants use `centralized_critic_q` postprocessing, right?

**Comment by Theohhhu** [comment]
Correct.

**Comment by XuehaiPan** [comment]
You may need to add

```python
{
    "normalize_actions": False,
    "clip_actions": False,
}
```

to

https://github.com/Replicable-MARL/MARLlib/blob/015dd257ccf196ba9cf59385f563e352f1a8ea75/marl/algos/core/IL/ddpg.py#L558-L569

Because you already unnormalized the action in `policy.model`.

See also ray-project/ray#24213.
